ksqldb-engine/src/main/java/io/confluent/ksql/engine/EngineExecutor.java

```diff
@@ -511,12 +506,36 @@ private boolean isSourceTableMaterializationEnabled() {
         .getBoolean(KsqlConfig.KSQL_SOURCE_TABLE_MATERIALIZATION_ENABLED);
   }
 
+  private Optional<SourceName> getSourceNameIfIsNotExists(final ConfiguredStatement<?> statement) {
+    if (statement.getStatement() instanceof CreateStream
+        && ((CreateStream) statement.getStatement()).isNotExists()) {
+      return Optional.ofNullable(((CreateStream) statement.getStatement()).getName());
+    } else if (statement.getStatement() instanceof CreateTable
+        && ((CreateTable) statement.getStatement()).isNotExists()) {
+      return Optional.ofNullable(((CreateTable) statement.getStatement()).getName());
+    } else if (statement.getStatement() instanceof CreateTableAsSelect
+        && ((CreateTableAsSelect) statement.getStatement()).isNotExists()) {
+      return Optional.ofNullable(((CreateAsSelect) statement.getStatement()).getName());
+    } else if (statement.getStatement() instanceof CreateStreamAsSelect
+        && ((CreateStreamAsSelect) statement.getStatement()).isNotExists()) {
+      return Optional.ofNullable(((CreateAsSelect) statement.getStatement()).getName());
+    }
+    return Optional.empty();
+  }
+
   // Known to be non-empty
+  // CHECKSTYLE_RULES.OFF: CyclomaticComplexity
   @SuppressWarnings("OptionalGetWithoutIsPresent")
-  KsqlPlan plan(final ConfiguredStatement<?> statement) {
+  Optional<KsqlPlan> plan(final ConfiguredStatement<?> statement) {
     try {
       throwOnNonExecutableStatement(statement);
 
+      final Optional<SourceName> sourceName = getSourceNameIfIsNotExists(statement);
+      if (sourceName.isPresent()
+          && engineContext.getMetaStore().getSource(sourceName.get()) != null) {
+        return Optional.empty();
+      }
+
```

Comment:
It would be more readable to have boolean method that just checks if the CREATE  statement can continue with the plan or not, like:
```
    if (!shouldGeneratePlanFor(statement)) { 
       return Optional.empty();
    }
```

This way we could handle any future statement that should not participate in the plan.